### ISSUE TITLE
Support a `RealmOptional` for single results

### ISSUE THREAD

**Issue by tmtrademarked** [realm/realm-java#5179]
#### Goal

As discussed in https://github.com/realm/realm-java/issues/4826, it is currently somewhat awkward to correctly handle the case of single item queries that might return null. When running `findFirst`, if the object doesn't exist yet, the query returns null - and you can't register a listener on a null object.

This is totally solvable today by using a RealmResults and `findAll` instead (with documentation that the size of the list is expected to be 0 or 1), but that's somewhat awkward code. It doesn't look or feel right.

#### Expected Results

A better solution would be to leverage something like Optional, where the query can return an object that holds either the actual value or `null`. This Optional could have a RealmListener applied to it which would allow the developer to receive callbacks when the item is created or updated. This allows code to preserve the look and feel of returning a single object while correctly leveraging query listeners.

**Comment by Zhuinden** [comment]
`RealmOptional` would definitely make the intention of `findFirstAsync()` (but maybe even `findFirst()`) much clearer, as it would be listening for a "single-element collection" that can later be updated, deleted, and a new element added which would also call the added change listener.

**Comment by cmelchior** [comment]
I do agree that the use case of `finding item X and only that` is one that isn't that well supported right now, and we did have quite a lot of talks about the semantics of `findFirst()` vs. `findAll()` . 

Personally, I think having something like a specific `RealmOptional<Class> obj  = realm.get(Class, id)` method is probably the best way to support this, since `findFirst` primary use case seems to be to lookup objects by their id. @tmtrademarked This was also your use case, right?

But in general having `RealmOptional` would definitely encode the intent much much better.

**Comment by tmtrademarked** [comment]
Yup, that was our use case - although it wasn't specifically by their primary key, but rather a different field. (Various internal reasons for that particular requirement, but we could probably re-work that without too much trouble)
